### PR TITLE
Barcode scanning: part IV

### LINKED ISSUE #1629: Barcode: review feedback
## Describe the solution you'd like 🎁
- [x] Implement `getBarcode` api call as per [this comment](https://github.com/openmsupply/open-msupply/pull/1564#discussion_r1170729405)
- [x] Remove reject in promise as per [this comment](https://github.com/openmsupply/open-msupply/pull/1564#discussion_r1178528032)
- [x] Remove masterlist parts from `AddFromScannerButton`
- [x] Reduce the queried fields as per [this comment](https://github.com/openmsupply/open-msupply/pull/1564#discussion_r1178583219)
- [ ] Change `value` to `barcode` or something?
- [x] tests for `parseBarcodeData`

### PR DESCRIPTION
**Pull request by mark-prins**
Fixes #1629 

# 👩🏻‍💻 What does this PR do? 
 
- [ ] Implement `getBarcode` api call as per [this comment](https://github.com/openmsupply/open-msupply/pull/1564#discussion_r1170729405)
- [ ] Remove reject in promise as per [this comment](https://github.com/openmsupply/open-msupply/pull/1564#discussion_r1178528032)
- [ ] Remove masterlist parts from `AddFromScannerButton`
- [ ] Reduce the queried fields as per [this comment](https://github.com/openmsupply/open-msupply/pull/1564#discussion_r1178583219)
- [ ] tests for `parseBarcodeData`

I've left `value` for the barcode value - couldn't cope with `barcode.barcode` sorry.

# 🧪 How has/should this change been tested? 

Tested with electron, and running queries manually.

You might find the barcoding examples in the [references](https://github.com/openmsupply/open-msupply/pull/1601#issuecomment-1526656151) helpful

## 📃 Documentation

- [ ] Barcoding

### CODE REVIEW

**Comment by andreievg** [comment]
> I've left value for the barcode value - couldn't cope with barcode.barcode sorry.

Oki, we can always add comments in database and graphQL schema

File: client/packages/common/src/utils/BarcodeScannerContext.test.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
+import { parseBarcodeData, parseResult } from './BarcodeScannerContext';
+
+describe('barcode parsing', () => {
```
Comment (andreievg) [inline client/packages/common/src/utils/BarcodeScannerContext.test.ts:3]: Thanks, together with the link you've provided in previous PR, this makes full sense !

File: client/packages/common/src/utils/BarcodeScannerContext.tsx
Comment (andreievg) [inline client/packages/common/src/utils/BarcodeScannerContext.tsx:1]: startScan race condition for timeout seems like a good use of Promise.race():

```
  const startScan = async () => {
    setIsScanning(true);

    const timeoutPromise = new Promise<undefined>((_, reject) =>
      setTimeout(reject, SCAN_TIMEOUT_IN_MS, 'Scan timed out')
    );

    const getBarcodePromise = () =>
      new Promise<string | undefined>(async (resolve, reject) => {
        if (hasElectronApi) {
          const { startBarcodeScan } = electronNativeAPI;
          await startBarcodeScan();

          electronNativeAPI.onBarcodeScan((_event, data) =>
            resolve(parseBarcodeData(data))
          );
        } else if (hasNativeBarcodeScanner) {
          // Check camera permission
          await BarcodeScanner.checkPermission({ force: true });
          // make background of WebView transparent
          BarcodeScanner.hideBackground();
          resolve((await BarcodeScanner.startScan()).content); // start scanning and wait for a result
        } else {
          reject('Cannot find scan api');
        }
      });

    let result: ScanResult = {};

    try {
      let barcode = await Promise.race([timeoutPromise, getBarcodePromise()]);
      result = parseResult(barcode);
    } catch (e) {
      error(t('error.unable-to-read-barcode'))();
      console.log(e);
    } finally {
      await stopScan();
    }

    return result;
  };
```

Btw I mentioned in previous PR, is this right: 
https://github.com/openmsupply/open-msupply/blob/8baa708b907db71463665bb1ec349729a2cbc51f/client/packages/common/src/utils/BarcodeScannerContext.tsx#L97

error only when not electron ?

also i think in above implementation it's easier to do overall stopScan, I noticed that stopBarcodeScan is missing in current implementation only when doing the suggestion

File: client/packages/system/src/Stock/Components/StockLineForm.tsx [hunk 2]
```diff
@@ -42,24 +43,29 @@ interface StockLineFormProps {
...
-    if (!!result.content) {
-      const { batch, content, expiryDate, gtin } = result;
-      const barcode = gtin ?? content;
-      const draft = {
-        barcode,
-        batch,
-        expiryDate,
-      };
+    try {
+      const result = await startScan();
```
Comment (andreievg) [inline client/packages/system/src/Stock/Components/StockLineForm.tsx:55]: hmm, does startScan actually throw an error ?

Comment (andreievg) [reply client/packages/system/src/Stock/Components/StockLineForm.tsx:55]: I guess BarcodeScannerPlugin might throw an error ¯\_(ツ)_/¯

**Review by andreievg — APPROVED** [review]
Thanks, looks good, I guess the extra changes can be done in another PR, not sure what happens if we don't stop scanning in startScan (single scan)

File: client/packages/common/src/utils/BarcodeScannerContext.tsx
Comment (mark-prins) [reply client/packages/common/src/utils/BarcodeScannerContext.tsx:1]: > error only when not electron ?

Yeah, there's a comment about that, which I'd forgotten about. 

```
  if the timeout has been hit then an error is raised
  by the electron implementation, and the snack is shown
  in that error handler, no need to duplicate
```

**Pull request merged**